### ISSUE TITLE
Groups scope exposes groups as uuids

### ISSUE THREAD

**Issue by dvv** [kanidm/kanidm#2471]
Hi!

When using vouch to empoy kanidm as oauth2 provider I get the following claim for groups scope:
```
  "groups": [
    "00000000-0000-0000-0000-000000000035",
    "00000000-0000-0000-0000-000000000036",
    "9a017bf0-942f-4a02-8af1-0af3c6a7783d",
    "487dedd9-8891-40a5-97cf-04c93595d60a",
    "2f5545bc-847f-4908-89e7-3cb5480e95a9",
    "298b613b-9fdc-4edf-b8a9-8158d7e795bb",
    "3ad56e95-d8f0-4cf4-aed6-c06b78134e65",
    "e2f791bc-bcc5-4b45-9ad9-b03bc0d73ac7"
  ]
```

I wonder if it's by design to use uuids instead of names?

TIA

**Comment by Firstyear** [comment]
Yes it is - this is because group names *might* change so it's safer to use the stable identifier. I have considered that we provider both uuid *and* spn though as I think that may be nicer. 

Saying this, you can also use custom claims as an alternative.

**Comment by dvv** [comment]
I see. However IMHO for the resource server namely the names are of more value -- similar to DNS which translates human-friendly names to opaque machine ids -- application may use different IdP, content of IdP may be flushed and recreated (rotating the uuids).

Couldn't you provide more info on how to expose group spn's and/or use custom claims? This would help me moving on directly.
TIA

**Comment by Firstyear** [comment]
Custom claims are in the next release/version of the server. 

I will add the spn support shortly.

**Comment by Firstyear** [comment]
> I see. However IMHO for the resource server namely the names are of more value -- similar to DNS which translates human-friendly names to opaque machine ids -- application may use different IdP, content of IdP may be flushed and recreated (rotating the uuids).

you should also be looking at this the othry way. If the UUID ever changes it implies strongly to the RS that the group in question is now not thesame and not present, so it may not be trusted. That's why we use stable IDs like this.

However, groups also are "less likely" to have their names changed (unlike a person) meaning that exposing the SPN here is not quite as bad.

**Comment by dvv** [comment]
That's perfectly valid. Thank you! Please feel free to close the issue whenever it suits.

**Comment by Firstyear** [comment]
I think in this case, it's okay to expose the groups by spn here. So I'll close it once I add that quickly.

**Comment by dvv** [comment]
That's great! Thank you, @Firstyear

**Comment by dvv** [comment]
Hi!

I wonder if we could add two settings to oauth2 app:
- control what to report for `groups` scope: `uuid`, `name`, `uuid_and_name`
- control how to report the group name: `full`, `short` (w/o domain name)

The rationale is that the the `groups` claim may grow and easily exceed the cookie allowed length, in case when extensive grouping used.

**Comment by dvv** [comment]
Personally, to me the following would close the case:

```diff
diff --git a/server/lib/src/idm/oauth2.rs b/server/lib/src/idm/oauth2.rs
index dde4a250d..aba0189dc 100644
--- a/server/lib/src/idm/oauth2.rs
+++ b/server/lib/src/idm/oauth2.rs
@@ -1361,7 +1361,7 @@ impl<'a> IdmServerProxyWriteTransaction<'a> {
             };

             let s_claims = s_claims_for_account(o2rs, &account, &scopes);
-            let extra_claims = extra_claims_for_account(&account, &o2rs.claim_map, &scopes);
+            let extra_claims = extra_claims_for_account(o2rs, &account, &o2rs.claim_map, &scopes);

             let oidc = OidcToken {
                 iss,
@@ -2157,7 +2157,7 @@ impl<'a> IdmServerProxyReadTransaction<'a> {
                 let iss = o2rs.iss.clone();

                 let s_claims = s_claims_for_account(o2rs, &account, &scopes);
-                let extra_claims = extra_claims_for_account(&account, &o2rs.claim_map, &scopes);
+                let extra_claims = extra_claims_for_account(o2rs, &account, &o2rs.claim_map, &scopes);
                 let exp = expiry.unix_timestamp();

                 // ==== good to generate response ====
@@ -2433,6 +2433,7 @@ fn s_claims_for_account(
 }

 fn extra_claims_for_account(
+    o2rs: &Oauth2RS,
     account: &Account,

     claim_map: &BTreeMap<Uuid, Vec<(String, ClaimValue)>>,
@@ -2477,7 +2478,12 @@ fn extra_claims_for_account(
                 .iter()
                 .flat_map(|x| {
                     let proto_group = x.to_proto();
-                    [proto_group.spn, proto_group.uuid]
+                    if o2rs.prefer_short_username {
+                        let (name, _domain) = &proto_group.spn.split_once('@').unwrap_or(("", &proto_group.spn));
+                        [name.to_string()]
+                    } else {
+                        [proto_group.spn, proto_group.uuid]
+                    }
                 })
                 .collect(),
         );
```

**Comment by Firstyear** [comment]
> The rationale is that the the `groups` claim may grow and easily exceed the cookie allowed length, in case when extensive grouping used.

You could hit that anyway, which is why we don't use cookies, but bearer tokens. Even with names in groups, you can hit this limit. 

That PR/suggestion has a bug - in the future we want to do trusts, which means we will have trusted groups that have spn only and no name. So you shouldn't be splitting on @ here because then a group like idm_all_accounts from a trust partner collides with idm_all_accounts on our system. This is the othry reason we use spn, to make it explicit about the group and it's source.

**Comment by dvv** [comment]
Oh, I do need means to make groups shortly named.
I understand your point.

We'll be (hopefully) running kanidm in place of the existing solution which reports short group names, and of course noone will reconfigure everything else for the new long named standard.

What I propose is to introduce a config option which controls what to pass as groups claim.
I'm very far from Rust and the "solution" above is what I finally managed to be compiled without too much intervention.

**Comment by Firstyear** [comment]
We can do what we did for usernames, and show name if present, or spn if not.

**Comment by Firstyear** [comment]
This is resolved, we now present groups as uuid, spn, and the mix for legacy support.
